### PR TITLE
Addition of tests for Buffer Property

### PR DESCRIPTION
**Pull request by Sakthi-SM**
**What does this Pull Request accomplish?**
  Added a test to check for support of int32 in buffer property.

**Why should this Pull Request be merged?**
This implements [Task 2330431](https://dev.azure.com/ni/DevCentral/_workitems/edit/2330431): Add or verify tests for Buffer property and related data types

### CODE REVIEW

**Review by subash-suresh — APPROVED** [review]

File: tests/test_buffer_property.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+"Contains a collection of pytest tests that validates the buffer property."
```
Comment (keer1111) [inline tests/test_buffer_property.py:1]: Is this intended to be a comment?

File: tests/test_buffer_property.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+            # Setting a valid input buffer size of type int32
+            task.in_stream.input_buf_size = 2000000000
+            assert task.in_stream.input_buf_size == 2000000000
+
+            # Setting a invalid input buffer size greater than int32
+            try:
+                task.in_stream.input_buf_size = 4000000000
+            except DaqError:
+                assert task.in_stream.input_buf_size == 2000000000
```
Comment (keer1111) [inline tests/test_buffer_property.py:34]: is there also validation for case in the left limit. like less than 0?

**Review by keer1111 — APPROVED** [review]

File: tests/test_buffer_property.py (hunk 2 above)
Comment (Sakthi-SM) [reply tests/test_buffer_property.py:34]: Setting a negative value to buffer size is not possible. So I think this covers all the corner cases

File: tests/test_buffer_property.py (hunk 1 above)
Comment (Sakthi-SM) [reply tests/test_buffer_property.py:1]: It is expected as per `ni-style-guide` to define what a python file does at the start of the file.

File: tests/test_buffer_property.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+            # Resetting input buffer size
+            del task.in_stream.input_buf_size
+            assert task.in_stream.input_buf_size == 10000
```
Comment (zhindes) [inline tests/test_buffer_property.py:38]: this value is device-specific and subject to change over time. could you query the default value before the first set/get test and then assert it matches here? if that works, it is slightly more resilient.

**Review by zhindes — CHANGES_REQUESTED** [review]

File: tests/test_buffer_property.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        # Reset the pseudorandom number generator with seed.
+        random.seed(seed)
+
+        with nidaqmx.Task() as task:
+            task.ai_channels.add_ai_voltage_chan(any_x_series_device.ai_physical_chans[0].name)
+            task.timing.samp_timing_type = SampleTimingType.SAMPLE_CLOCK
+
+            # Setting a valid input buffer size of type int32
+            task.in_stream.input_buf_size = 2000000000
+            assert task.in_stream.input_buf_size == 2000000000
```
Comment (bkeryan) [inline tests/test_buffer_property.py:28]: I assume this test doesn't actually allocate a 2 billion sample buffer because it doesn't commit the task, it only verifies the task.

File: tests/test_buffer_property.py (hunk 3 above)
Comment (Sakthi-SM) [reply tests/test_buffer_property.py]: Updated as per the suggestion

**Review by bkeryan — APPROVED** [review]

File: tests/test_buffer_property.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+from nidaqmx.errors import DaqError
+from nidaqmx.tests.helpers import generate_random_seed
+from nidaqmx.tests.test_read_write import TestDAQmxIOBase
+
+
+class TestBufferProperty(TestDAQmxIOBase):
+    """Contains a collection of pytest tests that validates the buffer property."""
+
+    @pytest.mark.parametrize("seed", [generate_random_seed()])
+    def test_int32_property(self, any_x_series_device, seed):
```
Comment (bkeryan) [inline tests/test_buffer_property.py:17]: I'm sorry, I forgot to bring up NI's regression testing conventions: https://dev.azure.com/ni/DevCentral/_wiki/wikis/AppCentral.wiki/20934/Regression-Test-Conventions

Using given/when/then and arrange/act/assert would likely require splitting this test case into multiple test cases.

File: tests/test_buffer_property.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+import nidaqmx
+from nidaqmx.constants import SampleTimingType
+from nidaqmx.errors import DaqError
+from nidaqmx.tests.helpers import generate_random_seed
+from nidaqmx.tests.test_read_write import TestDAQmxIOBase
+
+
+class TestBufferProperty(TestDAQmxIOBase):
+    """Contains a collection of pytest tests that validates the buffer property."""
```
Comment (bkeryan) [inline tests/test_buffer_property.py:14]: Remove the test class. Test functions should be sufficient.

**Review by bkeryan — CHANGES_REQUESTED** [review]

**Review by zhindes — APPROVED** [review]
I'm happy once Brad's last couple of issues are resolved

File: tests/test_buffer_property.py [hunk 7]
```diff
@@ -1,39 +1,41 @@
...
+        task.timing.samp_timing_type = SampleTimingType.SAMPLE_CLOCK
 
-    @pytest.mark.parametrize("seed", [generate_random_seed()])
-    def test_int32_property(self, any_x_series_device, seed):
-        """Test for validating int32 attributes in buffer."""
-        # Reset the pseudorandom number generator with seed.
-        random.seed(seed)
+        # Setting a valid input buffer size of type int32
+        task.in_stream.input_buf_size = 2000000000
+        assert task.in_stream.input_buf_size == 2000000000
```
Comment (bkeryan) [inline tests/test_buffer_property.py:15]: Leave a blank line between the act and assert sections.
```suggestion
        # Setting a valid input buffer size of type int32
        task.in_stream.input_buf_size = 2000000000

        assert task.in_stream.input_buf_size == 2000000000
```

File: tests/test_buffer_property.py [hunk 8]
```diff
@@ -1,39 +1,41 @@
...
+    """Test for validating int32 attributes in buffer."""
+    with nidaqmx.Task() as task:
+        task.ai_channels.add_ai_voltage_chan(any_x_series_device.ai_physical_chans[0].name)
+        task.timing.samp_timing_type = SampleTimingType.SAMPLE_CLOCK
+        default_buffer_size = task.in_stream.input_buf_size
+        task.in_stream.input_buf_size = 2000000000
+
+        # Resetting input buffer size
+        del task.in_stream.input_buf_size
+        assert task.in_stream.input_buf_size == default_buffer_size
```
Comment (bkeryan) [inline tests/test_buffer_property.py:41]: Leave a blank line between the act and assert sections.

```suggestion
        # Resetting input buffer size
        del task.in_stream.input_buf_size

        assert task.in_stream.input_buf_size == default_buffer_size
```

File: tests/test_buffer_property.py [hunk 9]
```diff
@@ -1,39 +1,41 @@
...
+    """Test for validating int32 attributes in buffer."""
+    with nidaqmx.Task() as task:
+        task.ai_channels.add_ai_voltage_chan(any_x_series_device.ai_physical_chans[0].name)
+        task.timing.samp_timing_type = SampleTimingType.SAMPLE_CLOCK
+
+        # Setting a invalid input buffer size greater than int32
+        try:
+            task.in_stream.input_buf_size = 4000000000
+        except DaqError:
             assert task.in_stream.input_buf_size == 2000000000
```
Comment (bkeryan) [inline tests/test_buffer_property.py]: Oops, there's a better way to do this:

```suggestion
        # Setting a invalid input buffer size greater than int32
        with pytest.raises(DaqError):
            task.in_stream.input_buf_size = 4000000000

        assert task.in_stream.input_buf_size == 2000000000
```

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: tests/test_buffer_property.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    """Test for validating int32 attributes in buffer."""
+    with nidaqmx.Task() as task:
+        task.ai_channels.add_ai_voltage_chan(any_x_series_device.ai_physical_chans[0].name)
+        task.timing.samp_timing_type = SampleTimingType.SAMPLE_CLOCK
+
+        # Setting a invalid input buffer size greater than int32
+        try:
+            task.in_stream.input_buf_size = 4000000000
+        except DaqError:
+            assert task.in_stream.input_buf_size == 2000000000
```
Comment (bkeryan) [inline tests/test_buffer_property.py:28]: This passes if no exception is raised. Use [pytest.raises](https://docs.pytest.org/en/4.6.x/reference.html#pytest-raises)

File: tests/test_buffer_property.py [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    with nidaqmx.Task() as task:
+        task.ai_channels.add_ai_voltage_chan(any_x_series_device.ai_physical_chans[0].name)
+        task.timing.samp_timing_type = SampleTimingType.SAMPLE_CLOCK
+
+        # Setting a invalid input buffer size greater than int32
+        with pytest.raises(DaqError):
+            task.in_stream.input_buf_size = 800000000000
+            task.in_stream.read()
+
+        assert task.in_stream.input_buf_size == 1136082944
```
[the remark this answers is not in the document]
Comment (bkeryan) [reply tests/test_buffer_property.py:32]: No. That value is what you get when you truncate 800000000000 to 32 bits.

```
PS C:\dev\nidaqmx-python> python
Python 3.9.13 (tags/v3.9.13:6de2ca5, May 17 2022, 16:36:42) [MSC v.1929 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> '0x%x' % 800000000000
'0xba43b74000'
>>> 0x43b74000
1136082944
```

NI IO Trace shows what's going on: `input_buf_size` is truncating 800000000000 to 1136082944 before calling the DAQmx C API, and then `read()` is raising a `DaqError` that is unrelated to the buffer size.
image

Please file a GitHub issue about the truncation and change the test to generate a different error. 

Suggestion:
- Set `task.out_stream.output_buf_size` to a reasonable value.
- This should throw a `DaqError` with error code `DAQmxErrors.ATTRIBUTE_NOT_SUPPORTED_IN_TASK_CONTEXT` (-200452).
- Use `as exc_info` and assert that `exc_info.value.error_code` matches the constant.
- Rename the test case accordingly (e.g. "test__buffer__set_unsupported_int32_property__throws_daqerror").

File: tests/test_buffer_property.py
Comment (bkeryan) [inline tests/test_buffer_property.py:1]: Rename to `tests/component/test_buffer_properties.py`

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: tests/test_buffer_property.py [hunk 12]
```diff
@@ -0,0 +1,46 @@
+"Contains a collection of pytest tests that validates the buffer property."
+import pytest
+
+import nidaqmx
+from nidaqmx.constants import SampleTimingType
+from nidaqmx.errors import DaqError
+
+
+def test__buffer__set_int32_property__value_is_set(any_x_series_device):
```
Comment (bkeryan) [inline tests/test_buffer_property.py:9]: Sorry, I've been overlooking these "given" clauses. "Given buffer" doesn't make sense. Rename the tests to `test__ai_task__...`

File: tests/component/test_buffer_property.py [hunk 13]
```diff
@@ -18,21 +19,22 @@ def test__buffer__set_int32_property__value_is_set(any_x_series_device):
...
-    """Test for validating int32 attributes in buffer."""
+def test__ai_task__set_valid_value_to_unsupported_property__unsupported_error_raised(
+    any_x_series_device,
+):
+    """Test for validating unsupported attribute in buffer."""
     with nidaqmx.Task() as task:
         task.ai_channels.add_ai_voltage_chan(any_x_series_device.ai_physical_chans[0].name)
         task.timing.samp_timing_type = SampleTimingType.SAMPLE_CLOCK
 
         # Setting a invalid input buffer size greater than int32
```
Comment (bkeryan) [inline tests/component/test_buffer_property.py:30]: Please delete this comment.

**Review by bkeryan — APPROVED** [review]

**Review by zhindes — CHANGES_REQUESTED** [review]

File: tests/component/test_buffer_property.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (Sakthi-SM) [reply tests/component/test_buffer_property.py]: Sorry Zach I think Brad's comment was deleted by mistake and I hope you are pointing to the `# Setting a invalid input buffer size greater than int32` comment.
